src/limb.rs

```diff
@@ -0,0 +1,277 @@
+                                -> Range {
+
+        debug_assert!(limbs_gte_limb(max_exclusive, min_inclusive));
+        debug_assert!(max_exclusive[max_exclusive.len() - 1] > 0);
+
+        Range {
+            min_inclusive: min_inclusive,
+            max_exclusive: max_exclusive
+        }
+    }
+
+    fn leading_zero_bits(&self) -> usize {
+        let most_significant_limb =
+            self.max_exclusive[self.max_exclusive.len() - 1];
+
+        for i in 0..(LIMB_BITS - 1) {
+            let bit = 1 << (LIMB_BITS - 1 - i);
+            if most_significant_limb & bit > 0 {
+                return i;
+            }
+        }
+
+        unreachable!();
+    }
+
+    /// Are these little-endian limbs within the range?
+    ///
+    /// *Not* constant-time.
+    pub fn are_limbs_within(&self, limbs: &[Limb]) -> bool {
+        assert_eq!(self.max_exclusive.len(), limbs.len());
+
+        let is_gte_min = limbs_gte_limb(limbs, self.min_inclusive);
+        let is_lt_max = limbs_less_than_limbs_constant_time(limbs, self.max_exclusive);
+
+        is_lt_max && is_gte_min
+    }
+
+    // Loosely based on NSA Guide Appendix B.2:
+    // "Key Pair Generation by Testing Candidates".
+    //
+    // TODO: DRY-up with `ec::suite_b::private_key::generate_private_key`?
+    //
+    /// Chooses a positive integer within the range and stores it into `dest`
+    /// encoded as limbs ordered least-significant-limb to most-significant-limb.
+    ///
+    /// This function is intended to be suitable for generating private keys.
+    pub fn sample_into_limbs(&self, dest: &mut [Limb], rng: &rand::SecureRandom)
+                             -> Result<(), ()> {
+
+        assert_eq!(self.max_exclusive.len(), dest.len());
+
+        // Ensure that we have at least a 50% chance of generating a matching
+        // number on each iteration; TODO: Either embed this in the type, or
+        // fix the algorithm so that it is unnecessary
+        debug_assert!(self.min_inclusive < (1 << LIMB_BITS - 1));
+
+        let bits_to_mask = self.leading_zero_bits();
+
+        // XXX: The value 100 was chosen to match OpenSSL due to uncertainty of
+        // what specific value would be better, but it seems bad to try 100 times.
+        for _ in 0..100 {
+            {
+                let mut dest_as_bytes = limbs_as_bytes_mut(dest);
+                try!(rng.fill(&mut dest_as_bytes));
+            }
+
+            if bits_to_mask > 0 {
+                let mask: Limb = (1 << (LIMB_BITS - bits_to_mask)) - 1;
+                dest[self.max_exclusive.len() - 1] &= mask;
+            }
+
+            if !self.are_limbs_within(&dest) {
+                continue;
+            }
+
+            return Ok(());
+        }
+
+        Err(())
+    }
+
+}
+
+#[allow(unsafe_code)]
+#[allow(non_snake_case)]
+#[doc(hidden)]
+#[no_mangle]
+pub unsafe extern fn GFp_rand_range_ex(dest: *mut Limb,
+                                       min_inclusive: Limb,
```

Comment:
Ditto, it seems `min_inclusive` is always 1, so it doesn't need to be a parameter.